src/ring_theory/prime_power.lean

```diff
@@ -0,0 +1,311 @@
+begin
+  let i := (normalized_factors r).card,
+  have hi : normalized_factors r = multiset.repeat (c 1) i,
+  { apply multiset.eq_repeat_of_mem,
+    intros b hb,
+    refine prime_dvd_eq_second_of_chain n hn c h₁ (λ r', h₂) (prime_of_normalized_factor b hb) hr
+      (dvd_of_mem_normalized_factors hb) },
+  have H : r = (c 1)^i,
+  { have := unique_factorization_monoid.normalized_factors_prod (ne_zero_of_dvd_ne_zero hq hr),
+    rw [associated_iff_eq, hi, multiset.prod_repeat] at this,
+    rw this },
+
+  refine ⟨⟨i, _⟩, H⟩,
+  have : (finset.univ.image (λ (m : fin (i + 1)), (c 1) ^ (m : ℕ))).card = i + 1,
+  { conv_rhs { rw [← finset.card_fin (i+1)] },
+    cases n, { contradiction },
+    rw finset.card_image_eq_iff_inj_on,
+    refine set.inj_on_of_injective (λ m m' h, fin.ext _) _,
+    refine pow_injective_of_not_unit (second_of_chain_not_is_unit (n+1) 1 dec_trivial c h₁) _ h,
+    exact irreducible.ne_zero (second_of_chain_is_irreducible _ hn c h₁ (λ r, h₂) hq) },
+
+  suffices H' : ∀ r ∈ (finset.univ.image (λ (m : fin (i + 1)), (c 1) ^ (m : ℕ))), r ≤ q,
+  { simp only [← nat.succ_le_iff, nat.succ_eq_add_one, ← this],
+    apply card_subset_divisors_le_length_of_chain n c (λ r', h₂) (finset.univ.image
+        (λ (m : fin (i + 1)), (c 1) ^ (m : ℕ))) H' },
+
+  simp only [finset.mem_image],
+  rintros r ⟨a, ha, rfl⟩,
+  refine dvd_trans _ hr,
+  use (c 1)^(i - a),
+  rw pow_mul_pow_sub (c 1),
+  { exact H },
+  { exact nat.succ_le_succ_iff.mp a.2 }
+end
+
+lemma eq_prime_pow_of_has_chain {q : associates M} (n : ℕ) (hn : n ≠ 0)
+  (c : fin (n + 1) → associates M) (h₁ : strict_mono c)
+  (h₂ : ∀ {r : associates M}, r ≤ q ↔ ∃ i, r = c i)
+  (hq : q ≠ 0) : q = (c 1)^n :=
+begin
+  obtain ⟨i, hi'⟩ := mem_chain_eq_pow_second_of_chain n hn c h₁ (λ r, h₂) (dvd_refl q) hq,
+  suffices : n ≤ i,
+  { rwa le_antisymm (nat.succ_le_succ_iff.mp i.prop) this at hi' },
+
+  refine nat.le_of_succ_le_succ _,
+  calc n + 1 = (finset.univ : finset (fin (n + 1))).card : (finset.card_fin _).symm
+         ... = (finset.univ.image c).card :
+    (finset.card_image_eq_iff_inj_on.mpr (h₁.injective.inj_on _)).symm
+         ... ≤ (finset.univ.image (λ (m : fin (i + 1)), (c 1)^(m : ℕ))).card :
+          finset.card_le_of_subset _
+         ... ≤ (finset.univ : finset (fin (i + 1))).card : finset.card_image_le
+         ... = i + 1 : finset.card_fin _,
+  intros r hr,
+  obtain ⟨j, -, rfl⟩ := finset.mem_image.1 hr,
+  have := h₂.2 ⟨j, rfl⟩,
+  rw hi' at this,
+  obtain ⟨u, hu, hu'⟩ := (dvd_prime_pow (show prime (c 1), from _) i).1 this,
+  refine finset.mem_image.mpr ⟨u, finset.mem_univ _, _⟩,
+  { rw associated_iff_eq at hu', rw [fin.coe_coe_of_lt (nat.lt_succ_of_le hu), hu'] },
+  { rw ← irreducible_iff_prime, exact second_of_chain_is_irreducible n hn c h₁ (λ r, h₂) hq, }
+end
+
+variables {N : Type*} [cancel_comm_monoid_with_zero N] [unique_factorization_monoid N]
+
+lemma image_prime_pow_dvd_of_monotone {m p : associates M} {n : associates N}
```

Comment:
these names need to be namespaced if they're going to be kept as if, imo; `image` of what? it's not clear so I'm not sure it's the naming convention